sdks/python/apache_beam/io/filebasedsource.py

```diff
@@ -308,3 +313,102 @@ def read(self, range_tracker):
 
   def default_output_coder(self):
     return self._file_based_source.default_output_coder()
+
+
+class _ExpandIntoRanges(DoFn):
+
+  def __init__(
+      self, splittable, compression_type, desired_bundle_size, min_bundle_size):
+    self._desired_bundle_size = desired_bundle_size
+    self._min_bundle_size = min_bundle_size
+    self._splittable = splittable
+    self._compression_type = compression_type
+
+  def process(self, element, *args, **kwargs):
+    match_results = FileSystems.match([element])
+    for metadata in match_results[0].metadata_list:
+      splittable = (
+          self._splittable and
+          _determine_spelittability_from_compression_type(
+              metadata.path, self._compression_type))
+
+      if not splittable:
+        yield (metadata, OffsetRange(
+            0, range_trackers.OffsetRangeTracker.OFFSET_INFINITY))
+      else:
+        for split in OffsetRange(
+            0, metadata.size_in_bytes).split(
+                self._desired_bundle_size, self._min_bundle_size):
+          yield (metadata, split)
+
+
+class _Reshard(PTransform):
+
+  def expand(self, pvalue):
+    return (pvalue
```

Comment:
This will break in the presence of non-trivial triggers/merging windows; at least test and throw an error in that case?